### PR TITLE
transform-type-modifier

### PR DESCRIPTION
**Pull request by amitaiporat**

### CODE REVIEW

File: lollipop/types.py [hunk 1]
```diff
@@ -1292,3 +1293,42 @@ def __repr__(self):
...
+
+    Example:
+
+        Point = namedtuple('Point', ['x', 'y'])
+
+        PointType = Transform(
+          Tuple(Integer(), Integer()),
+          load=lambda values: Point(values[0], values[1]),
+          dump=lambda point: [point.x, point.y],
+        )
```
Comment (maximkulkin) [inline lollipop/types.py]: Please add example on how serialized/deserialized data will look like (since in this case it is not obvious)

File: tests/test_types.py [hunk 2]
```diff
@@ -1864,3 +1864,37 @@ def test_dumping_passes_context_to_inner_type_dump(self):
         context = object()
         DumpOnly(inner_type).dump('foo', context)
         assert inner_type.dump_context == context
+
```
Comment (maximkulkin) [inline tests/test_types.py]: Two blank lines between top-level declarations please

File: tests/test_types.py [hunk 3]
```diff
@@ -1864,3 +1864,37 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+        transform = Transform(
+                              Tuple([String(), String()]),
+                              load_modifier=lambda values: Type(values[0], values[1]),
+                              dump_modifier=lambda type1: [type1.foo, type1.bar],
+                             )
+        assert transform.dump(Type(foo = 'foo', bar = 'bar')) == \
+               ['foo', 'bar']
+        assert transform.load(['foo', 'bar']) == \
+                Type(foo='foo', bar='bar')
+
```
Comment (maximkulkin) [inline tests/test_types.py]: Remove extra empty lines

File: tests/test_types.py [hunk 4]
```diff
@@ -1864,3 +1864,37 @@ def test_dumping_passes_context_to_inner_type_dump(self):
         context = object()
         DumpOnly(inner_type).dump('foo', context)
         assert inner_type.dump_context == context
+
+class TestTransform:
+    def test_transform_simple_transform(self):
```
Comment (maximkulkin) [inline tests/test_types.py]: Please break tests to test each transformation separately.

File: tests/test_types.py [hunk 5]
```diff
@@ -1864,3 +1864,37 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+        inner_type = SpyType()
+        transform = Transform(inner_type,
+                              load_modifier=lambda foo: 'foo',
```
Comment (maximkulkin) [inline tests/test_types.py]: This test does not check which value will be passed to modifier. Also, it does not check which value will be passed to inner type. All this checks should all be separate tests.

File: tests/test_types.py [hunk 6]
```diff
@@ -3,9 +3,9 @@
 import datetime
 from lollipop.compat import OrderedDict
 from lollipop.types import MISSING, ValidationError, Type, Any, String, \
-    Number, Integer, Float, Boolean, DateTime, Date, Time, OneOf, List, Dict, \
-    Field, AttributeField, MethodField, FunctionField, Constant, Object, \
-    Optional, LoadOnly, DumpOnly, type_name_hint, dict_value_hint
+    Number, Integer, Float, Boolean, DateTime, Date, Time, Tuple, OneOf, List, \
+    Dict, Field, AttributeField, MethodField, FunctionField, ConstantField, Object, \
```
Comment (maximkulkin) [inline tests/test_types.py]: Please do not make unnecessary changes (e.g. ConstantField, type_name_hint)

File: lollipop/types.py [hunk 7]
```diff
@@ -413,36 +414,36 @@ def load(self, data, *args, **kwargs):
...
+                result.append(item_type.load(item, *args, **kwargs))
             except ValidationError as ve:
                 errors_builder.add_errors({idx: ve.messages})
         errors_builder.raise_errors()
 
         return super(Tuple, self).load(result, *args, **kwargs)
 
-    def dump(self, value, *args, **kwargs):
-        if value is MISSING or value is None:
+    def dump(self, data, *args, **kwargs):
```
Comment (maximkulkin) [inline lollipop/types.py]: It's a common convention to call value for `dump()` method a `value` and value for `load()` method - `data`.

File: lollipop/types.py [hunk 8]
```diff
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
```
Comment (maximkulkin) [inline lollipop/types.py]: We do not need that.

File: lollipop/types.py [hunk 9]
```diff
@@ -1292,3 +1293,42 @@ def __repr__(self):
...
+          dump=lambda point: [point.x, point.y],
+        )
+
+    :param Type inner_type: Data type.
+    :param load_modifier: Specifies loaded format for object.
+    :param dump_modifier: Specifies dump format for object.
+
+    Tuple([String(), Integer(), Boolean()]).load(['foo', 123, False])
+    """
+    def __init__(self, inner_type, load_modifier=None, dump_modifier=None):
```
Comment (maximkulkin) [inline lollipop/types.py]: There is no way to supply `pre_load` and `post_dump` modifiers.

File: lollipop/types.py [hunk 10]
```diff
@@ -1292,3 +1293,42 @@ def __repr__(self):
...
+        super(Transform, self).__init__()
+        self.inner_type = inner_type
+        self.load_modifier = load_modifier
+        self.dump_modifier = dump_modifier
+
+    def post_load(self, data, *args, **kwargs):
```
Comment (maximkulkin) [inline lollipop/types.py]: These are redundant. Just rename `load_modifier` property to `post_load` and it will work exactly the same with less code. Same applies to `dump_modifier` and `pre_dump`.

**Comment by maximkulkin** [comment]
Also, please be more descriptive with commit messages. Here is a good guideline summary: http://chris.beams.io/posts/git-commit/

File: lollipop/types.py [hunk 11]
```diff
@@ -1292,3 +1293,42 @@ def __repr__(self):
             klass=self.__class__.__name__,
             inner_type=repr(self.inner_type),
         )
+
+
+class Transform(Type):
```
Comment (maximkulkin) [inline lollipop/types.py]: Commit-wise, this should be a separate commit named something like "Add Transform modifier type", not a part of commit titled "minor errors in types.py fixed"

File: lollipop/types.py (hunk 1 above)
Comment (amitaiporat) [inline lollipop/types.py]: fixed.

File: tests/test_types.py (hunk 2 above)
Comment (amitaiporat) [inline tests/test_types.py]: addressed

File: lollipop/types.py (hunk 1 above)
Comment (amitaiporat) [inline lollipop/types.py]: done

**Comment by amitaiporat** [comment]
I address initial review comments. Thanks for the helpful tips

File: lollipop/types.py [hunk 12]
```diff
@@ -421,19 +420,19 @@ def load(self, data, *args, **kwargs):
 
         return super(Tuple, self).load(result, *args, **kwargs)
 
-    def dump(self, data, *args, **kwargs):
```
Comment (maximkulkin) [inline lollipop/types.py]: This content does not correspond to commit message. Please extract it as a separate commit

File: docs/quickstart.rst [hunk 13]
```diff
@@ -125,10 +125,10 @@ If you want more control on your data, you can specify additional validators: ::
 If you just need to validate date and not interested in result, you can use
 :func:`~Type.validate()` method: ::
 
-    print UserType.load({"email": "wasa"})
```
Comment (maximkulkin) [inline docs/quickstart.rst]: Please change commit message to something like "Fix validate example in quickstart documentation"

File: lollipop/types.py [hunk 14]
```diff
@@ -434,15 +434,15 @@ def dump(self, value, *args, **kwargs):
...
+                result.append(item_type.dump(item, *args, **kwargs))
             except ValidationError as ve:
                 errors_builder.add_errors({idx: ve.messages})
         errors_builder.raise_errors()
 
         return super(Tuple, self).dump(result, *args, **kwargs)
 
     def __repr__(self):
-        return '<{klass} of {item_types}>'.format(
+        return '<{klass} of {item_type}>'.format(
```
Comment (maximkulkin) [inline lollipop/types.py]: Looks like there is a bug again: you reference value with "item_type" (which is incorrect since it's multiple types) and then you pass the value as "item_types"

File: lollipop/types.py [hunk 15]
```diff
@@ -1355,3 +1355,43 @@ def __repr__(self):
...
+                 pre_dump=None, post_dump=None):
+        super(Transform, self).__init__()
+        self.inner_type = inner_type
+        self.pre_load = pre_load or identity
+        self.post_load = post_load or identity
+        self.pre_dump = pre_dump or identity
+        self.post_dump = post_dump or identity
+
+    def load(self, data, *args, **kwargs):
+        return self.post_load(self.inner_type.load(self.pre_load(data), *args, **kwargs))
```
Comment (maximkulkin) [inline lollipop/types.py]: You still need to account for cases when pre/post_load functions need to receive an extra "context" argument.

File: lollipop/types.py [hunk 16]
```diff
@@ -1355,3 +1355,43 @@ def __repr__(self):
...
+        )
+
+        PointType.dump((Point(x=1, y=2)))
+        # => [1,2]
+
+        PointType.load([1,2])
+        # => Point(x=1, y=2)
+
+    :param Type inner_type: Data type.
+    :param load_modifier: Specifies loaded format for object.
```
Comment (maximkulkin) [inline lollipop/types.py]: Please update parameter names and descriptions.

File: tests/test_types.py [hunk 17]
```diff
@@ -2002,3 +2002,40 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+        transformed = transform.dump('foo')
+        assert inner_type.dumped == 'bar'
+
+    def test_post_dump_calls_inner_type_dump_with_result_of_pre_dump(self):
+        inner_type = SpyType()
+        transform = Transform(inner_type, post_dump=lambda foo: foo + 'bar')
+        transformed = transform.dump('foo')
+        assert transform.dump('foo') == 'foobar'
+
+    def test_transform_passes_context_to_inner_type_dump(self):
```
Comment (maximkulkin) [inline tests/test_types.py]: You also need 4 more tests to test that all callback function can optionally receive a context argument.

File: tests/test_types.py [hunk 18]
```diff
@@ -2002,3 +2002,40 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+        inner_type = SpyType()
+        context = object()
+        transform = Transform(inner_type,
+                              post_load = lambda foo: 'foo',
```
Comment (maximkulkin) [inline tests/test_types.py]: Why do you specify these particular arguments here and left the other ones. Either do not specify any of arguments, or specify all of them (I prefer not specifying any arguments)

File: tests/test_types.py [hunk 19]
```diff
@@ -2002,3 +2002,40 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+
+
+class TestTransform:
+    def test_loading_calls_inner_type_load_with_result_of_pre_load(self):
+        inner_type = SpyType()
+        transform = Transform(inner_type, pre_load=lambda foo: 'bar')
+        transformed = transform.load('foo')
+        assert inner_type.loaded == 'bar'
+
+    def test_post_load_calls_value_from_inner_type_load(self):
```
Comment (maximkulkin) [inline tests/test_types.py]: The test name does not make sense to me. Why post load should call something (since it `should be called`).
`test_loading_calls_post_load_with_result_of_inner_type_load`

File: tests/test_types.py [hunk 20]
```diff
@@ -2002,3 +2002,90 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+                self.pre_loaded = data
+                return data
+        callbacks = Callbacks()
+        Transform(SpyType(), pre_load=callbacks.pre_load).load('foo')
+        assert callbacks.pre_loaded == 'foo'
+
+    def test_dump_calls_pre_dump_with_original_value(self):
+        class Callbacks():
+            def pre_dump(self, data):
+                self.pre_dump = data
```
Comment (maximkulkin) [inline tests/test_types.py]: Please change property name to `pre_dumped`, because otherwise you are overwriting the callback method itself.

File: tests/test_types.py [hunk 21]
```diff
@@ -2002,3 +2002,90 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+    def test_loading_calls_pre_load_with_original_value(self):
+        class Callbacks():
+            def pre_load(self, data):
+                self.pre_loaded = data
+                return data
+        callbacks = Callbacks()
+        Transform(SpyType(), pre_load=callbacks.pre_load).load('foo')
+        assert callbacks.pre_loaded == 'foo'
+
+    def test_dump_calls_pre_dump_with_original_value(self):
```
Comment (maximkulkin) [inline tests/test_types.py]: Fix naming according to convention and put it next to other dump method tests (like the first test in dumping tests section).

File: tests/test_types.py [hunk 22]
```diff
@@ -2002,3 +2002,90 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+        transform = Transform(inner_type, pre_load=lambda foo: 'bar')
+        transformed = transform.load('foo')
+        assert inner_type.loaded == 'bar'
+
+    def test_loading_calls_post_load_with_result_of_inner_type_load(self):
+        inner_type = SpyType()
+        transform = Transform(inner_type, post_load=lambda foo: foo + 'bar')
+        transformed = transform.load('foo')
+        assert inner_type.loaded == 'foo'
+        assert transform.load('foo') == 'foobar'
```
Comment (maximkulkin) [inline tests/test_types.py]: You do not need to call `load` again since you have already captured return value. How about this:
```python
result = transform.load('foo')
assert result == 'foobar'
assert inner_type.loaded == 'foo'
```

File: tests/test_types.py [hunk 23]
```diff
@@ -2002,3 +2002,90 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+
+    def test_dumping_calls_inner_type_dump_with_result_of_pre_dump(self):
+        inner_type = SpyType()
+        transform = Transform(inner_type, pre_dump=lambda foo: 'bar')
+        transformed = transform.dump('foo')
```
Comment (maximkulkin) [inline tests/test_types.py]: No need to capture value if you're not using it. It makes less visual noise and makes tests easier to read without being distracted by unnecessary detail.

File: tests/test_types.py [hunk 24]
```diff
@@ -2002,3 +2002,90 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+        context = object()
+        transform = Transform(inner_type,
+                              pre_load=lambda foo, context: [foo, context])
+        assert transform.load('foo', context) == ['foo', context]
+
+    def test_transform_passes_context_to_post_load(self):
+        inner_type = SpyType()
+        context = object()
+        transform = Transform(inner_type,
+                              post_load=lambda foo, context:[foo,context])
```
Comment (maximkulkin) [inline tests/test_types.py]: Inconsistent spacing: in previous test you had space after colon and space after comma, but in this one there are no spaces. Convention is to put space after colon and comma.

File: tests/test_types.py [hunk 25]
```diff
@@ -2002,3 +2002,90 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+        context = object()
+        transform = Transform(inner_type,
+                              pre_dump=lambda foo, context: [foo, context])
+        assert transform.dump('foo', context) == ['foo', context]
+
+    def test_transform_passes_context_to_post_dump(self):
+        inner_type = SpyType()
+        context = object()
+        transform = Transform(inner_type,
+                              post_dump=lambda foo, context:[foo,context])
```
Comment (maximkulkin) [inline tests/test_types.py]: Inconsistent spacing.

File: tests/test_types.py [hunk 26]
```diff
@@ -2002,3 +2002,90 @@ def test_dumping_passes_context_to_inner_type_dump(self):
...
+    def test_dumping_calls_inner_type_dump_with_result_of_pre_dump(self):
+        inner_type = SpyType()
+        transform = Transform(inner_type, pre_dump=lambda foo: 'bar')
+        transformed = transform.dump('foo')
+        assert inner_type.dumped == 'bar'
+
+    def test_dumping_calls_post_dump_with_result_of_inner_type_dump(self):
+        inner_type = SpyType()
+        transform = Transform(inner_type, post_dump=lambda foo: foo + 'bar')
+        transformed = transform.dump('foo')
```
Comment (maximkulkin) [inline tests/test_types.py]: This line is not necessary: neither result nor side effects are checked.
